extensions/amp-fit-text/1.0/fit-text.js

```diff
@@ -0,0 +1,152 @@
+/**
+ * Copyright 2020 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as Preact from '../../../src/preact';
+import * as styles from './fit-text.css';
+import {px, setStyle, setStyles} from '../../../src/style';
+import {useCallback, useLayoutEffect, useRef} from '../../../src/preact';
+
+const {LINE_HEIGHT_EM_} = styles;
+
+/**
+ * @param {!JsonObject} props
+ * @return {PreactDef.Renderable}
+ */
+export function FitText(props) {
+  const {
+    'children': children,
+    'minFontSize': minFontSize = 6,
+    'maxFontSize': maxFontSize = 72,
+    'width': width = props['style']['width'] || '100%',
+    'height': height = props['style']['height'] || '100%',
+    ...rest
+  } = props;
+  const contentRef = useRef(null);
+  const measurerRef = useRef(null);
+
+  const resize = useCallback(
+    (maxHeight, maxWidth) => {
+      if (!measurerRef.current) {
+        return;
+      }
+      const fontSize = calculateFontSize(
+        measurerRef.current,
+        Number(maxHeight),
+        Number(maxWidth),
+        minFontSize,
+        maxFontSize
+      );
+      getOverflowStyle(measurerRef.current, Number(maxHeight), fontSize);
+    },
+    [maxFontSize, minFontSize]
+  );
+
+  // Here and below, useLayoutEffect is used so intermediary font sizes
+  // during resizing are resolved before the component visually updates.
+  // Font size should readjust when container resizes.
+  useLayoutEffect(() => {
+    const node = contentRef.current;
+    if (!node) {
+      return;
+    }
+    const observer = new ResizeObserver((entries) => {
+      const last = entries[entries.length - 1];
+      const {height: maxHeight, width: maxWidth} = last['contentRect'];
+      resize(maxHeight, maxWidth);
+    });
+    observer.observe(node);
```

Comment:
This covers the parent size changes, but not the content changes. E.g. if the updated `children` are passed with more text, then font-size should get smaller to accommodate. We may either have to do to resize observers, one for `contentRef` and one for `measurerRef`. Or maybe just one for `measurerRef` will suffice. I also commented on the maxHeight issue below for measurerRef. All of these cases, btw, ideally would be storybook cases as well to track regressions.

Reply:
I'm not sure how we can track this in storybook since the component is forced to render (as if on first load) every time you change a knob. Is there a way to change state in Storybook and look for reactionary render rather than render caused by Storybook?

Reply:
The secondary `useLayoutEffect` [discussed here is](https://github.com/ampproject/amphtml/pull/28169#discussion_r424493242) for managing content changes.